Add NFT tutorial (#305)
tendermint/starport: docs/04 Use cases/03_nft/03_nft.md

# Create NFTs on the EVM with the Cosmos SDK (ERC721)

> Non-Fungible Token or in short NFT are Tokens that are non-divisible and live on the blockchain. Those Token can represent a variety of things, from Artworks, to Houses, to Ownership of Cars or digital Items in games. The usage is from individuals to companies or institutions that are leverage the power of NFTs. 

Reviewer: ```suggestion
Non-Fungible Token or in short NFT are tokens that are non-divisible and live on the blockchain. Those tokens can represent a variety of things, from pieces of artwork, to houses, to ownership of cars or digital Items in games. The usage is from individuals to companies or institutions that are leverage the power of NFTs. 
```

      host: "127.0.0.1",
      port: 8545,
      network_id: "*" // Match any network id
    }
  }
```

For the compiler we will use the following settings in this tutorial in the same file

```javascript
  // Configure your compilers
  compilers: {
    solc: {
      version: "0.4.24",    // Fetch exact version from solc-bin (default: truffle's version)
      // docker: true,        // Use "0.5.1" you've installed locally with docker (default: false)
      // settings: {          // See the solidity docs for advice about optimization and evmVersion
      //  optimizer: {
      //    enabled: false,
      //    runs: 200
      //  },
      //  evmVersion: "byzantium"
      // }
    },
```

This will be the connection our rest-server uses. In order to setup the rest server on your Cosmos SDK blockchain, the following command will get it started on the right port:

`ethappcli rest-server --laddr "tcp://localhost:8545" --trace --chain-id ethapp-1 --unlock-key me`

## Smart Contract Code

From here our setup is complete, we can now jump into the smart contract code. Since we are working with OpenZeppelin reference implementation, we need to install their components.

```bash

> npm i zeppelin-solidity

Reviewer 2: Shouldn't we add the `-g` option like for `truffle` ?

Reviewer 2: And if not, it should be a dev dependency

```

> In our `contract` directory we add the solidity code for the NFT ERC721 Token that we would like to address, create the `contract/MyToken.sol` file and add

Reviewer: ```suggestion
In our `contracts` directory we add the solidity code for the NFT ERC721 Token that we would like to address, create the `contracts/MyToken.sol` file and add
```

```solidity
pragma solidity ^0.4.24;

import "../node_modules/zeppelin-solidity/contracts/token/ERC721/ERC721Token.sol";

contract MyNFT is ERC721Token {
    constructor (string _name, string _symbol) public
        ERC721Token(_name, _symbol)
    {
    }

    function mintUniqueTokenTo(
        address _to,
        uint256 _tokenId,
        string  _tokenURI
    ) public
    {
        super._mint(_to, _tokenId);
        super._setTokenURI(_tokenId, _tokenURI);
    }
}
```

in the `migrations` directory we add the file for the deployment, `2_deploy_contract.js`.

```javascript
const MyNFT = artifacts.require("./MyNFT.sol");

module.exports = async function(deployer) {
  await deployer.deploy(MyNFT, "MyNFT", "MyNFT")
  const erc721 = await MyNFT.deployed()
};
```

> From this we an already launch and deploy our contract. When the blockchain is running and our rest-server is setup, we can now migrate using truffle. 

Reviewer 2: ```suggestion
From this, we launch and deploy our contract. When the blockchain is running and our rest-server is setup, we can now migrate using truffle. 
```
